### DISCUSSION TITLE
Saving authors in a many to many table

### DISCUSSION THREAD

**Discussion by jcc5018** [filamentphp/filament#14112] [Help]
### Package

Panel builder

### Package Version

v3.2.110

### How can we help you?

I am attempting to create an article system that allows for multiple contributors to an article. 

`
Schema::create('articles', function (Blueprint $table) {
            $table->bigIncrements('id');
            $table->string('title')->index();
            $table->string('slug')->unique();
            $table->longText('content');
            $table->string('excerpt')->nullable();
            $table->string('status')->default('draft')->nullable();
            $
        });

 Schema::create('article_user', function (Blueprint $table) {
            $table->id();
            $table->foreignId('article_id')->constrained();

            $table->foreignId('user_id')->nullable()->constrained();

            $table->boolean('can_post')->default(1);
            $table->foreignId('group_id')->nullable()->constrained();
            $table->string('post_as')
                  ->nullable();`
                 
So the original creator of the article should be filled in automatically into the article_user table. But then I also need the post_as attribute to allow users a choice of how they want their name to appear (username, full name, anonymous, for instance)

Group_id is present in the event an article is created on behalf of a group or organization. 

If another author contributes to the article, then they will also have a choice how they wish their name to appear.

So, the user_id aspect should be based on the user_id , but how do I also handle the extra pivot data? 

I dont see much in regard to many-to many relationships other than using a relationship manager which is not really appropriate for this situation.  

This is what i have so far, though i am probably missing the relationship.  And I don't yet have the user_id defined. But I'm not sure how to get the user_id into the article_user table when this is dealing with the ArticleResource. 

` Select::make('post_as')
                                          ->options(function () {
                                              return Article::POST_AS;
                                          })
                                          ->default('username')
                                          ->label('Post As'),
`

Article Model

`  const POST_AS
        = [
            'username'  => 'Username',
            'fullname'  => 'Real Name',
            'anonymous' => 'Anonymous',
            'group'     => 'Group Name',
            'both'      => 'Group and Name',
        ];
        
public function users(): BelongsToMany
    {
        return $this->belongsToMany(User::class)
                    ->withPivot('can_post', 'group_id', 'post_as');
    }`
![image](https://github.com/user-attachments/assets/97248b7b-a613-450c-8c69-d977bd4277a1)

Any help would be greatly appreciated.  I tried github copilot and it  suggest something about the "saved callback of the form, [where I ]  can manually sync the relationship:"

But I don't really see anything about dealing with "saved callbacks" in the docs either, so I'm not sure if the AI is correct in this approach, or referencing something not applicable to filament.  I'm sure its possible, but its a bit over my head on how it is supposed to work.

**Comment by micraux** [comment]
Hi,

First, please use 3 backticks followed by _`php`_ to highlight your code:
````
```php
   // Your code here
```
````

ARTICLES -> STATUS
As you use `default('Draft')`, I think you could omit `nullable()`...

ARTICLE_USER -> POST_AS
Take a look at [enums](https://filamentphp.com/docs/3.x/support/enums), that should be useful for your use case.

You may need an extra ARTICLE_USER -> CREATOR (boolean) in order to know who created the article.

ARTICLE_USER -> CAN_POST
I can't understand this: is it a kind of policy, eg. user can post article? If so, it has (IMHO) nothing to do here...

Needing or not a RelationManager depends on what you want to achieve:
- Do you want to auto-populate the list of contributors whenever one modify the article ?
- Do you need to operate on that list to add/remove a contributor for an article?

BTW, you are not describing any problem: what have you tried so far, and/or what issue are you facing?

**Reply by jcc5018** [reply]
Hi, first, thanks for replying. I have tried various things for the code blocks in various forums that use mark down, and I struggle to get it to parse correctly. In this instance I believe i used the code block button in the editor, which only has one tic. But i've also tried 3 and 4 tics at times as well and the system still doesnt format the code blocks correctly. So I just gave up in this instance.  Not sure what the issue is, but I try. 

I removed nullable from status. --you are right, probably not needed

I've examined enums in the past. When i started this project enums were not supported or overly complicated, so My const system seems to work ok for this case. 

article_user-> creator might be worth considering.   Ultimately my goal is to set this up for potential wiki type editing on certain articles where multiple people can edit. 

I did describe the problem and What i have tried. Though maybe i should have labeled it better. 

I'll restate it.

Problem:
"So the original creator of the article should be filled in automatically into the article_user table. But then I also need the post_as attribute to allow users a choice of how they want their name to appear (username, full name, anonymous, for instance)"

"So, the user_id aspect should be based on the user_id , but how do I also handle the extra pivot data?"

Simplified: On the Article resource, How do I send the user_id to a secondary table and how do I handle the pivot data on that table.

WHAT I HAVE TRIED:
This is what i have so far, though i am probably missing the relationship. And I don't yet have the user_id defined. But I'm not sure how to get the user_id into the article_user table when this is dealing with the ArticleResource.

`Select::make('post_as') ->options(function () { return Article::POST_AS; }) ->default('username') ->label('Post As'),`


ADDITIONAL NOTES:

If I was just using a user_id on the article table, I'd have a hidden::make('user_id)    set up.  But that user_id needs to go to the article_user table along with how they want their name to appear. 

There is no need for a relationship manager. Who ever creates the article automatically sets an entry in the pivot table.  This should default to the logged in user_id. The only option they need in the set up is how they want their name to appear.

Then if article type is wiki or allows contributors... anyone who provides an update will also create an entry in the pivot table based on their logged in user_id, again with the option to choose how their name should appear. 

There will be no "selection of users" or adding or removing contributors.  If someone has content being displayed, they are listed as contributors. 

Which does lead me to the question of defining which content is from which person, so I might need to rethink my article structure as well. 

I designed this system a long time before actually trying to build it. So i forget what I intended the can_post attribute for. I think to prevent certain users to do any further editing of an article or something of that nature in the event they spam it or something. But I might get rid of that later. 


You know of any prebuilt wiki type systems for laravel/ filament?

**Comment by micraux** [comment]
Hi,

> 
> If I was just using a user_id on the article table, I'd have a hidden::make('user_id) set up. But that user_id needs to go to the article_user table along with how they want their name to appear.
> 
> There is no need for a relationship manager. Who ever creates the article automatically sets an entry in the pivot table. This should default to the logged in user_id. The only option they need in the set up is how they want their name to appear.

This is exactly the purpose of Relation Managers!

You would just add a SELECT field for POST_AS and the RM would put it in the pivot table: [Create related records](https://filamentphp.com/docs/3.x/panels/resources/relation-managers#creating-with-pivot-attributes)

If you don't want to use RMs, you can simply [customize the creation process](https://filamentphp.com/docs/3.x/panels/resources/creating-records#customizing-the-creation-process).

> Which does lead me to the question of defining which content is from which person, so I might need to rethink my article structure as well.

Yes, if you want to keep a track of "who did what", you surely need to rethink the whole structure.

Take a look here: [wiki for Laravel](https://laravel-news.com/bookstack-documentation-wiki-software)
